File: ts/src/samchon/protocol/external/ExternalSystemArray.ts
/// <reference path="../../API.ts" />

/// <reference path="../EntityCollection.ts" />

namespace samchon.protocol.external
{
	export abstract class ExternalSystemArray
		extends EntityArrayCollection<ExternalSystem>
		implements IProtocol
	{
		/* ---------------------------------------------------------
			CONSTRUCTORS
		--------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();
		}

		/* ---------------------------------------------------------
			ACCESSORS
		--------------------------------------------------------- */
		public hasRole(key: string): boolean
		{
			for (let i: number = 0; i < this.size(); i++)
				for (let j: number = 0; j < this.at(i).size(); j++)
					if (this.at(i).at(j).key() == key)
						return true;

			return false;
		}

		public getRole(key: string): ExternalSystemRole
		{
			for (let i: number = 0; i < this.size(); i++)
				for (let j: number = 0; j < this.at(i).size(); j++)
					if (this.at(i).at(j).key() == key)
						return this.at(i).at(j);

			throw new std.OutOfRange("No role with such name.");
		}

		/* ---------------------------------------------------------
			MESSAGE CHAIN
		--------------------------------------------------------- */
		public sendData(invoke: Invoke): void
		{
			for (let i: number = 0; i < this.size(); i++)
				this.at(i).sendData(invoke);
		}

		public replyData(invoke: Invoke): void
		{
			invoke.apply(this);
		}

		/* ---------------------------------------------------------
			EXPORTERS
		--------------------------------------------------------- */
		public TAG(): string
		{
			return "systemArray";
		}
		public CHILD_TAG(): string
		{
			return "system";
		}
	}
}

File: cpp/samchon/protocol/external/ExternalSystemArray.cpp
#include <samchon/protocol/external/ExternalSystemArray.hpp>

using namespace std;
using namespace samchon;
using namespace samchon::library;
using namespace samchon::protocol;
using namespace samchon::protocol::external;

/* =========================================================
	CONSTRUCTORS
		- DEFAULT
		- FACTORY METHODS FOR CHILDREN
============================================================
	DEFAULT
--------------------------------------------------------- */
ExternalSystemArray::ExternalSystemArray()
	: super()
{
}
ExternalSystemArray::~ExternalSystemArray()
{
}

/* ---------------------------------------------------------
	FACTORY METHODS FOR CHILDREN
--------------------------------------------------------- */
//void ExternalSystemArray::addClient(shared_ptr<ClientDriver> driver)
//{
//	ExternalSystem *system = createExternalClient(driver);
//	if (system == nullptr)
//		return;
//
//	system->communicator = driver;
//
//	emplace_back(system);
//	driver->listen(system);
//}
//auto ExternalSystemArray::createChild(shared_ptr<XML> xml) -> ExternalSystem*
//{
//	return this->createExternalServer(xml);
//}

/* =========================================================
	NETWORK
		- SERVER AND CLIENT
		- MESSAGE CHAIN
============================================================
	SERVER AND CLIENT
--------------------------------------------------------- */
//void ExternalSystemArray::connect()
//{
//	for (size_t i = 0; i < size(); i++)
//	{
//		auto external_server = std::dynamic_pointer_cast<ExternalServer>(this->at(i));
//		external_server->connect();
//	}
//}

void ExternalSystemArray::sendData(shared_ptr<Invoke> invoke)
{
	for (size_t i = 0; i < size(); i++)
		at(i)->sendData(invoke);
}